multicluster/controllers/multicluster/node_controller.go

```diff
@@ -73,65 +78,215 @@ func NewNodeReconciler(
 
 func (r *NodeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	klog.V(2).InfoS("Reconciling Node", "node", req.Name)
+	if !r.initialized {
+		if err := r.initialize(); err != nil {
+			return ctrl.Result{}, err
+		}
+		r.initialized = true
+	}
 	gw := &mcsv1alpha1.Gateway{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      req.Name,
 			Namespace: r.namespace,
 		},
 	}
-	// When the Node is annotated with 'multicluster.antrea.io/gateway=true' as a Gateway:
-	//   - Delete the Gateway if the Node is deleted
-	//   - Update the Gateway if Node's InternalIP or GatewayIP is updated
-	//   - Create a new Gateway if there is no existing Gateway
+
 	node := &corev1.Node{}
 	if err := r.Client.Get(ctx, req.NamespacedName, node); err != nil {
 		if apierrors.IsNotFound(err) {
-			err := r.Client.Delete(ctx, gw, &client.DeleteOptions{})
-			return ctrl.Result{}, client.IgnoreNotFound(err)
+			if r.activeGateway == req.Name {
+				if err := r.recreateActiveGateway(ctx, gw); err != nil {
+					return ctrl.Result{}, err
+				}
+				return ctrl.Result{}, nil
+			}
 		}
 		klog.ErrorS(err, "Failed to get Node", "node", req.Name)
-		return ctrl.Result{}, err
+		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
-	_, isGW := node.Annotations[common.GatewayAnnotation]
-	var err error
-	gwNamespacedName := types.NamespacedName{
-		Name:      node.Name,
-		Namespace: r.namespace,
+	_, hasGWAnnotation := node.Annotations[common.GatewayAnnotation]
```

Comment:
It is still hard for me to read the code and check all cases. I feel it should be much easier to handle isActiveGateway and noActiveGateway separately, and if needed we can define a separate func for common code. Here is what I get:

```
	stillGatewayNode := false
	node := &corev1.Node{}
	if err := r.Client.Get(ctx, namespacedName, node); err != nil {
		if !apierrors.IsNotFound(err) {
			klog.ErrorS(err, "Failed to get Node", "node", req.Name)
			return ctrl.Result{}, err
		}
	} else {
		_, hasGWAnnotation := node.Annotations[common.GatewayAnnotation]
		stillGatewayNode = hasGWAnnotation
	}

	if stillGatewayNode {
		r.gatewayCandidates[req.Name] = true
	} else {
		delete(gatewayCandidates, req.Name)
	}

	noActiveGateway := r.activeGateway == ""
	isActiveGateway := r.activeGateway == req.Name

	if isActiveGateway {
		if stillGatewayNode {
			if gw := r.getValidGateway(node); gw != nil {
				r.updateActiveGateway()
				return
			}
		}
		recreateActiveGateway() // As it called only once now, maybe we can remove the func and change to delete (here) + create (merge with the noActiveGateway case).
		return
	}

	if noActiveGateway {
		if !stillGatewayNode {
			return ctrl.Result{}, nil
		}

		if r.getValidGateway(node) != nil {
			r.createGateway()
		}
		return
	}
```

The common code can be like:

```
func (r *NodeReconciler) getValidGateway(node *corev1.Node) *mcsv1alpha1.Gateway {
	nodeIsReady := isReadyNode(node)
	if nodeIsReady {
		return nil
	}
	
	internalIP, gwIP, err := r.getGatawayNodeIP(node)
	if err != nil {
		klog.ErrorS(err, "There is no valid Gateway IP for Node", "node", node.Name)
		return nil
	}
	
	return &mcsv1alpha1.Gateway{
		ObjectMeta: metav1.ObjectMeta{
			Name:      req.Name,
			Namespace: r.namespace,
		},
		InternalIP: internalIP,
	    GatewayIP: gwIP,
	}
}
```

What you think?